taichi/analysis/build_cfg.cpp

```diff
@@ -52,18 +64,69 @@ class CFGBuilder : public IRVisitor {
     return node;
   }
 
+  /**
+   * Structure:
+   *
+   * block {
+   *   node {
+   *     ...
+   *   } -> (node_loop_begin), (node_loop_end);
```

Comment:
What is the relationship between `node`, `node_loop_begin` and `node_loop_begin`? Also, why are the later two quoted with `()`? (Is it that they are not in the graph?)

Reply:
The nodes with `()` are not handled in this function (`void visit(ContinueStmt *stmt)` in this case). See the definition of `node_loop_begin` and `node_loop_end` in the docstring of `void visit_loop(Block *body, CFGNode *before_loop, bool is_while_true)`.

The node `(next node)` is not yet created, and probably doesn't exist. `(next node)` just shows that the next node will be at that place in the block.

Reply:
Oh, that shouldn't be `(node_loop_end)`...

Also, `node_loop_begin` is already created, so maybe it's clearer to remove the parentheses here.

Reply:
Ah OK.

> The nodes with () are not handled in this function

Maybe put such info to the beginning of the documentation. It will also be great to explain how to interpret the "structures", as well as giving a short example :-)

For example, it's not really obvious what this means:

>  -> node_loop_begin, (the next node after the loop);

I guess it means `continue` brings the control flow back to `node_loop_begin`, but how exactly?

Reply:
Right, but I think maybe it's too long to put them in the docstring. 

> I guess it means `continue` brings the control flow back to `node_loop_begin`, but how exactly?

We have a vector `continues_in_current_loop`, and we insert the control-flow edges in `visit_loop()`.

Reply:
>  I think maybe it's too long to put them in the docstring.

Which part would be too long? CFG is a rather important piece in the IR optimization passes, so it would make sense to document this part well. In addition, it's not the full CFG algorithm that we need to document, just some words to explain how the structure should be interpreted:

As an example:

* `node_a -> node_b, node_c`: does this mean `a` connects to both `b` and `c`?
* `node -> node_loop_begin, (next node) if !is_while_true;`: Does this mean `a` connects to either `node_loop_begin` or `(next node)`; or `a` always connects to `node_loop_begin`, and could connect to `(next node)` if `!is_while_true`?

Reply:
I see! I'll add some more documentations.